### ISSUE TITLE
Auto Exit

### ISSUE THREAD

**Issue by zeelrupapara** [zeelrupapara/trade-engine#9]
# Explanation of Auto-Exit Logic Concepts

### 1. Fixed Take-Profit and Stop-Loss

* **Meaning:** Set fixed price thresholds at trade entry to close a position automatically when profit or loss reaches a certain percentage.
* **Example:**

  * Stop-loss = -2% (if price drops 2% from entry, close trade)
  * Take-profit = +4% (if price rises 4%, close trade)
* **Pros:** Simple to implement
* **Cons:** Fixed levels ignore changes in market volatility (e.g., price swings may trigger stop-loss too early).

---

### 2. Trailing Exits

* **Meaning:** Move the stop-loss price as the market price moves favorably to lock in profits while allowing the position to run.
* **Example:** If price rises, move stop-loss up by a fixed amount or percentage behind the current price, trailing it like a leash. If price reverses and hits stop-loss, exit trade.
* **Pros:** Captures more profits during trends
* **Cons:** Slightly complex to implement because stop price updates dynamically.

---

### 3. Indicator-Based Exit

* **Meaning:** Use technical indicators to signal when to exit a position.
* **Example:**

  * Exit long if RSI (Relative Strength Index) rises above 70 (overbought) or price crosses below a moving average (e.g., 200-day MA).
  * Exit short if RSI < 30 (oversold) or price crosses above moving average.
* **Literature example:** “Sell when price drops below 200-day MA and RSI < 30.”
* **Pros:** Adaptive to market conditions
* **Cons:** Requires indicator calculation and interpretation.

---

### 4. Rolling Averages or Statistical Thresholds

* **Meaning:** Use short-term moving averages or recent price-volume data to detect reversals and exit.
* **Example:**

  * If price closes below 20-bar moving average, exit long.
  * Calculate average return of last N ticks; if PnL dips below negative of that average (indicating quick reversal), exit.
* **Pros:** Uses recent price action for dynamic exits
* **Cons:** Requires calculation of rolling stats.

---

### 5. Profit Targets via PnL

* **Meaning:** Set profit-taking targets based on multiples of risk or volatility.
* **Example:**

  * If current Profit and Loss (PnL) >= 1.5 times risk taken, take profit.
  * Or exit if profit exceeds 99th percentile of expected move (based on volatility).
* **Pros:** Dynamically linked to actual trade performance
* **Cons:** Needs PnL tracking and statistical data.

---

### 6. ATR-Based Volatility Exit

* **Meaning:** Use Average True Range (ATR) to set exits based on volatility.
* **Example:**

  * Exit profitable trade if price moves by more than 2 ATR from entry price or if ATR signals a reversal.
* **Pros:** Adjusts for volatility, avoids premature exit in volatile markets
* **Cons:** Needs ATR calculation.

---

# Mermaid Flowchart Algorithm for Auto-Exit Logic

Since Mermaid doesn't allow brackets in node text, I replaced `( )` with words or dashes:

```mermaid
flowchart TD
    Start --> CheckPosition

    CheckPosition -->|No position open| End
    CheckPosition -->|Position open| EvaluateExit

    EvaluateExit --> FixedExit
    EvaluateExit --> TrailingExit
    EvaluateExit --> IndicatorExit
    EvaluateExit --> RollingAverageExit
    EvaluateExit --> ProfitTargetExit
    EvaluateExit --> ATRExit

    FixedExit -->|Price hits stop-loss or take-profit| ClosePosition
    FixedExit -->|No| Continue

    TrailingExit -->|Price triggers trailing stop| ClosePosition
    TrailingExit -->|No| Continue

    IndicatorExit -->|Indicator signals exit| ClosePosition
    IndicatorExit -->|No| Continue

    RollingAverageExit -->|Price below rolling average or PnL threshold| ClosePosition
    RollingAverageExit -->|No| Continue

    ProfitTargetExit -->|PnL exceeds target| ClosePosition
    ProfitTargetExit -->|No| Continue

    ATRExit -->|ATR exit condition met| ClosePosition
    ATRExit -->|No| Continue

    Continue --> End
    ClosePosition --> End
```

---

# How to use this flowchart in code

* **Start**: Your system checks if there is an open position.
* If yes, it evaluates multiple exit conditions in parallel or sequence.
* Each exit method checks its own condition.
* If any exit condition triggers, it closes the position.
* Otherwise, it continues monitoring until next tick or price update.
